local-macros/src/lib.rs

```diff
@@ -0,0 +1,88 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::Data::Struct;
+use syn::{parse_macro_input, DeriveInput, Fields, FieldsNamed};
+
+#[proc_macro_attribute]
+pub fn ast_type(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let item_for_parsing = item.clone();
+    let DeriveInput {
+        ident: ast_type_name,
+        data,
+        ..
+    } = parse_macro_input!(item_for_parsing);
+
+    let node_interface_implementation = match data {
+        Struct(struct_) => {
+            let first_field_name = match struct_.fields {
+                Fields::Named(FieldsNamed { named, .. }) => named
+                    .iter()
+                    .nth(0)
+                    .expect("Expected at least one struct field")
+                    .ident
+                    .clone(),
+                _ => panic!("Expected named fields"),
+            };
+
+            quote! {
+                impl crate::NodeInterface for #ast_type_name {
+                    fn kind(&self) -> crate::SyntaxKind {
+                        self.#first_field_name.kind()
```

Comment:
This is assuming the convention where the first struct field is where the nested `BaseNode` lives so we're just forwarding these `NodeInterface` methods on to that field